File: App/lib/models/diagnostic_center.dart
class DiagnosticCenter {
  final String id;
  final String name;
  final String address;
  final double latitude;
  final String suburb;
  final double longitude;
  final double rating;
  final int reviewsCount;
  final String crowdLevel;
  final int estimatedWaitMinutes;
  final String phone;
  final String timing;
  final bool certified;
  final Map<String, double> testPrices;
  bool isFavorite;

  // Google Places only returns name, rating, review count and location.
  // Everything price/wait related is a heuristic estimate, not real data.
  // These flags let the UI label it honestly and the ranker ignore it until
  // a real price/wait source is wired in.
  final bool priceIsEstimated;
  final bool waitIsEstimated;

  DiagnosticCenter({
    required this.id,
    required this.name,
    required this.address,
    required this.suburb,
    required this.latitude,
    required this.longitude,
    required this.rating,
    required this.reviewsCount,
    required this.crowdLevel,
    required this.estimatedWaitMinutes,
    required this.phone,
    required this.timing,
    required this.certified,
    required this.testPrices,
    this.isFavorite = false,
    this.priceIsEstimated = true,
    this.waitIsEstimated = true,
  });
}

class EnhancedCenter {
  final DiagnosticCenter center;
  final double distanceKm;
  final double testPrice;
  final int matchScore;
  final bool priceIsEstimated;

  EnhancedCenter({
    required this.center,
    required this.distanceKm,
    required this.testPrice,
    required this.matchScore,
    this.priceIsEstimated = true,
  });
}

File: App/app/src/main/java/com/example/data/model/DiagnosticCenter.kt
package com.example.data.model

import androidx.room.Entity
import androidx.room.PrimaryKey

@Entity(tableName = "diagnostic_centers")
data class DiagnosticCenter(
    @PrimaryKey val id: Int,
    val name: String,
    val address: String,
    val suburb: String,
    val latitude: Double,
    val longitude: Double,
    val rating: Float,
    val reviewsCount: Int,
    val crowdLevel: String, // "Low", "Moderate", "High"
    val crowdPercentage: Int, // 10% - 100%
    val estimatedWaitMinutes: Int,
    val phone: String,
    val timing: String,
    val certified: Boolean,
    val testsJson: String, // e.g. {"CBC": 25, "MRI Brain": 210, ...}
    val isFavorite: Boolean = false,
    // Which mode produced this row: "DIAGNOSTIC" | "PHARMACY" | "DOCTOR".
    // Results are scoped by this so pharmacy/doctor searches never mix with labs.
    val category: String = "DIAGNOSTIC",
    // Pharmacy only: inferred size tier ("Small", "Mid-sized", "Brand / chain",
    // "Hospital pharmacy"). Inferred from name/reviews/hospital proximity — never
    // real inventory data, so the UI labels it "inferred".
    val tierLabel: String = "",
    // Doctor only: the specialty this result was matched against.
    val specialtyLabel: String = "",
    // The place's own website, straight from Google Places. This is the only real
    // route to a centre's published price list — no API exposes test prices.
    val websiteUri: String = ""
) {
    // Helper to get tests as a map
    fun getTestPrices(): Map<String, Double> {
        val map = mutableMapOf<String, Double>()
        try {
            val json = org.json.JSONObject(testsJson)
            val keys = json.keys()
            while (keys.hasNext()) {
                val key = keys.next()
                map[key] = json.optDouble(key)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return map
    }
}
